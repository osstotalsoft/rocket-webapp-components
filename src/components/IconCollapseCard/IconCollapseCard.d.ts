import { CardProps } from '@material-ui/core/Card';
import { CardContentProps } from '@material-ui/core/CardContent';
import { CardHeaderProps } from '@material-ui/core/CardHeader';
import { AccordionProps } from '@material-ui/core/Accordion';
import { AccordionDetailsProps } from '@material-ui/core/AccordionDetails';
import { AccordionSummaryProps } from '@material-ui/core/AccordionSummary';
export interface Props {
    /**
     * Icon to be displayed. This property is mandatory
     */
    icon: Object;
    /**
    * @default "theme"
    * The color of the icon that will be displayed
    */
    iconColor?: "orange" |
                "green" |
                "red" |
                "blue" |
                "purple" |
                "rose" |
                "theme"
    /**
    * The title to be displayed. This property can be a string or a component
    */
    title?: string | React.ReactNode;
    /**
    * The content of the card. This property can be a string or a component
    * @Warning 
    * if you want to add a component in the title make sure its onClick handler calls event.stopPropagation() to prevent events firing for overlapping components 
    */
    content?: React.ReactNode;
    /**
    * The footer to be displayed. This property can be a string or a component
    */
    footer?: React.ReactNode;
    /**
    * If set to true, the background will be transparent
    */
    plain?: boolean;
    /**
    * Override or extend the styles applied to the content of the card
    */
    customCardContentClass?: string;
    /**
    * The summary of the card
    */
    summary?: React.ReactNode;
    /**
    * If set to true, the card will be expanded
    */
    expanded?: boolean;
    /**
    * If set to true, the card can be expanded
    */
    canExpand?: boolean;
    /**
    * Default value for expanded property. If it's not provided will be set to false
    */
    defaultExpanded?: boolean;
    /**
     * @param {object} event The event source of the callback.
     * @param {boolean} expanded The `expanded` state of the card.
     */
    onToggle?: (event: React.SyntheticEvent, eventExpanded: boolean) => void;
    /**
     * Demos:
     * 
     * - https://bit.dev/totalsoft_oss/react-mui/icon-collapse-card
     * 
     *  Take a look over the code for further information:https://github.com/osstotalsoft/rocket-webapp-components/blob/master/components/IconCollapseCard/IconCollapseCard.js 
     */
}

export type IconCollapseCardProps = Props | CardProps | CardContentProps | CardHeaderProps | AccordionProps | AccordionDetailsProps | AccordionSummaryProps

export default function IconCollapseCard(props: IconCollapseCardProps): JSX.Element;