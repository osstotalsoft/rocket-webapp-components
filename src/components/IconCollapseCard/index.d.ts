export { default } from './IconCollapseCard';
export * from './IconCollapseCard';