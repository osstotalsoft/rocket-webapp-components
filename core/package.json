{
  "engines": {
    "node": ">=10"
  },
  "scripts": {
    "prebuild": "rimraf build",
    "build": "npm run build:cjs &&  npm run build:copy-files",
    "build:cjs": "babel src -d build --copy-files",
    "build:copy-files": "node ./scripts/copy-files.js",
    "bit-build": "bit -v && bit import && bit build"
  },
  "devDependencies": {
    "@babel/cli": "^7.8.0",
    "@babel/core": "^7.7.7",
    "@babel/node": "^7.7.7",
    "@babel/plugin-proposal-decorators": "^7.8.3",
    "@babel/preset-env": "^7.7.7",
    "@babel/preset-react": "^7.8.0",
    "babel-eslint": "^10.0.3",
    "cross-env": "^7.0.2",
    "eslint": "^6.8.0",
    "eslint-config-prettier": "^6.7.0",
    "eslint-config-standard": "^14.1.0",
    "eslint-config-standard-react": "^9.2.0",
    "eslint-plugin-import": "^2.18.2",
    "eslint-plugin-node": "^11.0.0",
    "eslint-plugin-prettier": "^3.1.1",
    "eslint-plugin-promise": "^4.2.1",
    "eslint-plugin-react": "^7.17.0",
    "eslint-plugin-standard": "^4.0.1",
    "gh-pages": "^2.2.0",
    "npm-run-all": "^4.1.5",
    "prettier": "^1.19.1",
    "react": "^16.13.1",
    "react-dom": "^16.13.1",
    "react-scripts": "^3.4.1",
    "rimraf": "^3.0.0",
    "url-search-params-polyfill": "^7.0.1"
  },
  "peerDependencies": {
    "react": "^16.13.1",
    "react-dom": "^16.13.1",
    "react-flags": "^0.1.18",
    "react-router-dom": "^5.1.2"
  },
  "sideEffects": false,
  "dependencies": {},
  "bit": {
    "env": {
      "compiler": "totalsoft_oss.react-mui/compilers/react@1.0.32"
    },
    "componentsDefaultDirectory": "components/{name}",
    "packageManager": "npm"
  }
}
