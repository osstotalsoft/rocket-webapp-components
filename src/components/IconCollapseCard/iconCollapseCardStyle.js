import styles from "../../assets/jss/styles.js";
import defaultTheme from "../themes/defaultTheme.js";

const iconCollapseCardStyle = theme => {
  const {
    card,
    cardHeader,
    orangeCardHeader,
    greenCardHeader,
    redCardHeader,
    blueCardHeader,
    purpleCardHeader,
    roseCardHeader,
    themeCardHeader,
    themeWithGradientCardHeader
  } = styles(theme);

  return {
    card,
    cardPlain: {
      background: "transparent",
      boxShadow: "none"
    },
    cardHeader: {
      ...cardHeader,
      float: "left"
    },
    orangeCardHeader,
    greenCardHeader,
    redCardHeader,
    blueCardHeader,
    purpleCardHeader,
    roseCardHeader,
    themeCardHeader,
    themeWithGradientCardHeader,
    cardContent: {
      padding: "15px 20px",
      position: "relative"
    },
    cardAvatar: {
      margin: "0px"
    },
    cardIcon: {
      paddingTop: "3px",
      paddingLeft: "3px",
      paddingRight: "3px",
      color: "#FFFFFF",
      width: "30px",
      height: "30px"
    },
    cardTitle: {
      ...(theme.header?.title || defaultTheme?.header?.title),
      display: "flex",
      marginBottom: "3px",
      width: "100%"
    },
    cardCategory: {
      color: "#999999",
      fontSize: "14px",
      fontWeight: "400",
      lineHeight: "1",
      marginBottom: "0"
    },
    cardFooter: {
      margin: "0 20px 10px",
      paddingTop: "10px",
      borderTop: "1px solid #eeeeee"
    },
    accordionRoot: {
      boxShadow: "none",
      backgroundColor: "transparent",
      padding: "0px"
    },
    expansionPanelDetails: {
      "@media (max-width: 480px)": {
        padding: "4px"
      }
    },
    content: {
      margin: "0px !Important",
      padding: "0px"
    },
    summaryRoot: {
      padding: "0px"
    }
  };
};

export default iconCollapseCardStyle;
