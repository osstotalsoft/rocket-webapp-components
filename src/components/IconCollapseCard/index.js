import IconCollapseCard from "./IconCollapseCard";
export default IconCollapseCard;