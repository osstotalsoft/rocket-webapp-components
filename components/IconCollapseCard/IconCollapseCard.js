import React, { useState, useCallback } from "react";
import cx from "classnames";
import PropTypes from "prop-types";
import {
  Card,
  CardContent,
  CardHeader,
  Accordion,
  AccordionDetails,
  AccordionSummary,
  makeStyles
} from "@material-ui/core";
import iconCollapseCardStyle from "./iconCollapseCardStyle";
import Typography from "../Typography";
import ExpandMoreIcon from "@material-ui/icons/ExpandMore";

const useStyles = makeStyles(iconCollapseCardStyle);

function IconCollapseCard({
  defaultExpanded,
  onToggle,
  expanded,
  title,
  content,
  iconColor,
  footer,
  plain,
  customCardContentClass,
  icon: Icon,
  summary,
  canExpand,
  className,
}) {
  const [localExpanded, setExpanded] = useState(defaultExpanded || false);
  const classes = useStyles();

  const handleToggle = useCallback(
    (_, eventExpanded) => {
      if (typeof onToggle !== "undefined") onToggle(_, eventExpanded);
      else {
        setExpanded(eventExpanded);
      }
    },
    [onToggle]
  );

  const exp = expanded || localExpanded;
  const canExp = typeof canExpand === "undefined" ? true : canExpand;

  const cardClasses =
    classes.card +
    " " +
    cx({
      [classes.cardPlain]: plain
    }) +
    " " +
    cx({
      [className]: className != null
    });

  const cardContentClasses =
    classes.cardContent +
    " " +
    cx({
      [customCardContentClass]: customCardContentClass !== undefined
    });

  return (
    <Card className={cardClasses}>
      <CardHeader
        classes={{
          root: classes.cardHeader + " " + classes[iconColor + "CardHeader"],
          avatar: classes.cardAvatar,
        }}
        avatar={<Icon className={classes.cardIcon} />}
      />
      <CardContent className={cardContentClasses}>
        {canExp ? (
          <Accordion
            onChange={handleToggle}
            disabled={!canExp}
            expanded={exp}
            className={classes.accordionRoot}
          >
            <AccordionSummary
              classes={{ content: classes.content, root: classes.summaryRoot }}
              expandIcon={<ExpandMoreIcon />}
            >
              {(exp || (!exp && !summary)) && (
                <Typography variant="subtitle1" className={classes.cardTitle}>
                  {title}
                </Typography>
              )}
              {!exp && summary && (
                <div style={{ width: "100%" }}>{summary}</div>
              )}
            </AccordionSummary>
            <AccordionDetails className={classes.expansionPanelDetails}>
              {content}
            </AccordionDetails>
          </Accordion>
        ) : (
          <Accordion
            onChange={handleToggle}
            expanded={false}
            className={classes.accordionRoot}
          >
            <AccordionSummary
              classes={{ content: classes.content, root: classes.summaryRoot }}
            >
              {(exp || (!exp && !summary)) && (
                <Typography variant="subtitle1" className={classes.cardTitle}>
                  {title}
                </Typography>
              )}
              {!exp && summary && (
                <div style={{ width: "100%" }}>{summary}</div>
              )}
            </AccordionSummary>
          </Accordion>
        )}
      </CardContent>
      {footer && <div className={classes.cardFooter}>{footer}</div>}
    </Card>
  );
}

IconCollapseCard.defaultProps = {
  iconColor: "theme",
};

IconCollapseCard.propTypes = {
  /**
   * Icon to be displayed. This property is mandatory
   */
  icon: PropTypes.object.isRequired,
  /**
   * The color of the icon that will be displayed
   */
  iconColor: PropTypes.oneOf([
    "orange",
    "green",
    "red",
    "blue",
    "purple",
    "rose",
    "theme",
    "themeWithGradient"
  ]),
  /**
   * The title to be displayed. This property can be a string or a component
   */
  title: PropTypes.node,
  /**
   * The content of the card. This property can be a string or a component
   */
  content: PropTypes.node,
  /**
   * The footer to be displayed. This property can be a string or a component
   */
  footer: PropTypes.node,
  /**
   * If set to true, the background will be transparent
   */
  plain: PropTypes.bool,
  /**
   * Override or extend the styles applied to the content of the card
   */
  customCardContentClass: PropTypes.string,
  /**
   * The summary of the card
   */
  summary: PropTypes.node,
  /**
   * If set to true, the card will be expanded
   */
  expanded: PropTypes.bool,
  /**
   * If set to true, the card can be expanded
   */
  canExpand: PropTypes.bool,
  /**
   * Default value for expanded property. If it's not provided will be set to false
   */
  defaultExpanded: PropTypes.bool,
  /**
   * @param {object} event The event source of the callback.
   * @param {boolean} expanded The `expanded` state of the card.
   */
  onToggle: PropTypes.func,
  /**
   * The className of card.
   */
  className: PropTypes.string,
};

export default IconCollapseCard;
