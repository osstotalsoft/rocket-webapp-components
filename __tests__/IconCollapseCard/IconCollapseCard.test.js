import React from "react";
import IconCollapseCard from "../../components/IconCollapseCard/IconCollapseCard";
import AssignmentLate from "@material-ui/icons/AssignmentLate";
import { mount } from 'enzyme';
import { Accordion } from "@material-ui/core";

describe("IconCollapseCard", () => {
  it("iconColor should be equal to default value", () => {
    const wrapper = mount(
      <IconCollapseCard
        icon={AssignmentLate}
        title={"Title card"}
        content={"Content card"}
        footer={"Footer card"}
      />
    );

    expect(wrapper.props().iconColor).toBe("theme")
  });

  it("if canExpand is undefined then Accordion is by default expandable and enabled", () => {
    const wrapper = mount(
      <IconCollapseCard
        icon={AssignmentLate}
        title={"Title card"}
        content={"Content card"}
        footer={"Footer card"}
      />
    );

    expect(wrapper.find(Accordion).props().disabled).toBe(false)
    expect(wrapper.find(Accordion).props().className).toContain('accordionRoot')
  });

  it("if canExpand is false then Accordion is by default not-expandable and disabled", () => {
    const wrapper = mount(
      <IconCollapseCard
        icon={AssignmentLate}
        title={"Title card"}
        content={"Content card"}
        footer={"Footer card"}
        canExpand={false}
      />
    );

    expect(wrapper.find(Accordion).props().expanded).toBe(false)
  });
});
