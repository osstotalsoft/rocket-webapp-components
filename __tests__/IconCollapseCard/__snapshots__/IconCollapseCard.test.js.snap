// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IconCollapseCard Snapshot 1`] = `
<div
  className="MuiPaper-root MuiCard-root makeStyles-card-1   MuiPaper-elevation1 MuiPaper-rounded"
>
  <div
    className="MuiCardHeader-root makeStyles-cardHeader-3 makeStyles-themeCardHeader-10"
  >
    <div
      className="MuiCardHeader-avatar makeStyles-cardAvatar-12"
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root makeStyles-cardIcon-13"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M19 3h-4.18C14.4 1.84 13.3 1 12 1c-1.3 0-2.4.84-2.82 2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zm-6 15h-2v-2h2v2zm0-4h-2V8h2v6zm-1-9c-.55 0-1-.45-1-1s.45-1 1-1 1 .45 1 1-.45 1-1 1z"
        />
      </svg>
    </div>
    <div
      className="MuiCardHeader-content"
    />
  </div>
  <div
    className="MuiCardContent-root makeStyles-cardContent-11 "
  >
    <div
      className="MuiPaper-root MuiAccordion-root makeStyles-accordionRoot-17 MuiAccordion-rounded MuiPaper-elevation1 MuiPaper-rounded"
    >
      <div
        aria-disabled={false}
        aria-expanded={false}
        className="MuiButtonBase-root MuiAccordionSummary-root makeStyles-summaryRoot-20"
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        role="button"
        tabIndex={0}
      >
        <div
          className="MuiAccordionSummary-content makeStyles-content-19"
        >
          <h6
            className="MuiTypography-root makeStyles-cardTitle-14 MuiTypography-subtitle1"
          >
            Title card
          </h6>
        </div>
        <div
          aria-disabled={false}
          aria-hidden={true}
          className="MuiButtonBase-root MuiIconButton-root MuiAccordionSummary-expandIcon MuiIconButton-edgeEnd"
          onBlur={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          role={null}
          tabIndex={null}
        >
          <span
            className="MuiIconButton-label"
          >
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
              />
            </svg>
          </span>
        </div>
      </div>
      <div
        className="MuiCollapse-root MuiCollapse-hidden"
        style={
          Object {
            "minHeight": "0px",
          }
        }
      >
        <div
          className="MuiCollapse-wrapper"
        >
          <div
            className="MuiCollapse-wrapperInner"
          >
            <div
              role="region"
            >
              <div
                className="MuiAccordionDetails-root makeStyles-expansionPanelDetails-18"
              >
                Content card
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className="makeStyles-cardFooter-16"
  >
    Footer card
  </div>
</div>
`;
